firmware/controllers/engine_cycle/high_pressure_fuel_pump.cpp

```diff
@@ -12,12 +12,170 @@
+	// correct itself rather quickly.
+	if (next_index >= engineConfiguration->hpfpCamLobes) {
+		next_index -= engineConfiguration->hpfpCamLobes;
+	}
+	m_lobe_index = next_index;
+
+	// Calculate impact of VVT
+	angle_t vvt = 0;
+	if (engineConfiguration->hpfpCam != HPFP_CAM_NONE) {
+		// TODO: Is the sign correct here?  + means ATDC?
+		vvt = engine->triggerCentral.getVVTPosition(
+			(engineConfiguration->hpfpCam - 1) / 2 & 1, // Bank
+			(engineConfiguration->hpfpCam - 1) & 1);    // Cam
+	}
+
+	return engineConfiguration->hpfpPeakPos + vvt +
+		next_index * 720 / engineConfiguration->hpfpCamLobes;
+}
+
+// As a percent of the full pump stroke
+float HpfpQuantity::calcFuelPercent(int rpm) {
+	float fuel_requested_cc_per_cycle =
+		ENGINE(injectionMass)[0] * (1.f / fuelDensity) * engineConfiguration->specs.cylindersCount; // TODO: handle multiple banks?
+	float fuel_requested_cc_per_lobe = fuel_requested_cc_per_cycle / engineConfiguration->hpfpCamLobes;
+	return 100.f *
+		fuel_requested_cc_per_lobe / engineConfiguration->hpfpPumpVolume +
+		interpolate3d(engineConfiguration->hpfpCompensation,
+			      engineConfiguration->hpfpCompensationLoadBins, fuel_requested_cc_per_lobe,
+			      engineConfiguration->hpfpCompensationRpmBins, rpm);
+}
+
+float HpfpQuantity::calcPI(int rpm, float calc_fuel_percent) {
+	m_pressureTarget = std::max<float>(
+		m_pressureTarget - (engineConfiguration->hpfpTargetDecay * 60. /
+				    (rpm * engineConfiguration->hpfpCamLobes)),
+		interpolate3d(engineConfiguration->hpfpTarget,
+			      engineConfiguration->hpfpTargetLoadBins, Sensor::get(SensorType::Map).Value, // TODO: allow other load axis, like we claim to
+			      engineConfiguration->hpfpTargetRpmBins, rpm));
+
+	float pressureError = m_pressureTarget - Sensor::get(SensorType::FuelPressureHigh).Value;
+
+	float p_control = pressureError * engineConfiguration->hpfpPidP;
+	float i_control = m_I_sum + pressureError * engineConfiguration->hpfpPidI;
+	calc_fuel_percent += p_control;
+	i_control = clampF(-calc_fuel_percent, i_control, 100.f - calc_fuel_percent);
+	m_I_sum = i_control;
+	return p_control + i_control;
+}
+
+angle_t HpfpQuantity::pumpAngleFuel(int rpm) {
+	// Math based on fuel requested
+	float fuel_requested_percent = calcFuelPercent(rpm);
+	
+	// Apply PI control
+	fuel_requested_percent += calcPI(rpm, fuel_requested_percent);
+	
+	// Convert to degrees
+	return interpolate2d(fuel_requested_percent,
+			     engineConfiguration->hpfpLobeProfileQuantityBins,
+			     engineConfiguration->hpfpLobeProfileAngle);
+}
+
+void HpfpActor::scheduleNextCycle() {
+#if EFI_UNIT_TEST
+	Engine *engine = this->engine;
+	EXPAND_Engine;
+#endif /* EFI_UNIT_TEST */
+
+	// TODO: All this math should be moved to a slower task/thread/callback/etc
+
+	angle_t lobe = m_lobe.findNextLobe();
+
+	// Pressure current/target calculation
+	int rpm = ENGINE(rpmCalculator).getRpm();
+
+	// What conditions can we not handle?
+	if (rpm < 10 ||
```

Comment:
i have a feeling that we have some magic constant for min possible RPM and it's something like 60? i could be wrong. anyway "10" looks really weird. maybe "< crankingRpm"?

Reply:
I'd like the HPFP logic to work when cranking, so not < crankingRPM.  60 (1/sec) is more reasonable than 10.  I didn't see the constant anywhere though.